csi.proto

```diff
@@ -1634,3 +1679,190 @@ message NodeExpandVolumeResponse {
+}
+
+message GroupControllerGetCapabilitiesResponse {
+  // All the capabilities that the group controller service supports.
+  // This field is OPTIONAL.
+  repeated GroupControllerServiceCapability capabilities = 1;
+}
+
+// Specifies a capability of the group controller service.
+message GroupControllerServiceCapability {
+  message RPC {
+    enum Type {
+      UNKNOWN = 0;
+
+      // Indicates that the group controller plugin supports
+      // creating and deleting a volume group snapshot.
+      CREATE_DELETE_VOLUME_GROUP_SNAPSHOT = 1
+      [(alpha_enum_value) = true];
+
+      // Indicates whether the group controller plugin supports
+      // creating a volume from an individual volume snapshot if the
+      // volume snapshot is part of a VolumeGroupSnapshot.
+      // Use cases: selective restore, advanced recovery, etc.
+      INDIVIDUAL_SNAPSHOT_RESTORE = 2 [(alpha_enum_value) = true];
+
+      // Indicates that the group controller plugin supports getting
+      // details of a volume group snapshot.
+      GET_VOLUME_GROUP_SNAPSHOT = 3 [(alpha_enum_value) = true];
+
+      // Indicates that the group controller plugin supports getting
+      // details of a list of volume group snapshots.
+      LIST_VOLUME_GROUP_SNAPSHOTS = 4 [(alpha_enum_value) = true];
+    }
+
+    Type type = 1;
+  }
+
+  oneof type {
+    // RPC that the controller supports.
+    RPC rpc = 1;
+  }
+}
+message CreateVolumeGroupSnapshotRequest {
+  option (alpha_message) = true;
+
+  // suggested name for a group snapshot (required for idempotent)
+  // This field is REQUIRED.
+  string name = 1;
+
+  // volume ids of the volumes to be snapshotted together.
+  // This field is REQUIRED.
+  repeated string volume_ids = 2;
+
+  // Secrets required by plugin to complete volume group snapshot
+  // creation request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 3 [(.csi.v1.csi_secret) = true];
+
+  // Plugin specific parameters passed in as opaque key-value pairs.
+  // This field is OPTIONAL. The Plugin is responsible for parsing and
+  // validating these parameters. COs will treat these as opaque.
+  map<string, string> parameters = 4;
+}
+
+message CreateVolumeGroupSnapshotResponse {
+  option (alpha_message) = true;
+
+  // Contains all attributes of the newly created group snapshot.
+  // This field is REQUIRED.
+  VolumeGroupSnapshot group_snapshot = 1;
+}
+
+message VolumeGroupSnapshot {
+  option (alpha_message) = true;
+
+  // The identifier for this group snapshot, generated by the plugin.
+  // This field is REQUIRED.
+  string group_snapshot_id = 1;
+
+  // A list of snapshots created. Snapshot is the same
+  // definition as Snapshot definition used in CSI.
+  // This field is OPTIONAL.
+  repeated Snapshot snapshots = 2;
+
+  // Indicates if a list of group snapshots are ready.
+  // This field is REQUIRED.
+  bool ready_to_use = 3;
```

Comment:
just to understand, this field turns `ready` only when all the snapshots in a group is ready or atleast one?

Reply:
I can't see what this field would be used for. Since only the individual snapshots can be cloned, all that matters is whether the individual snapshots are ready_to_use.

If in the future we develop a way to clone the whole group at once, and the group's ready_to_use can be different from the logical AND of all the member's ready_to_use field (not sure how that could be, but I'm open minded about it) then we could add a ready_to_use field to the group snapshot at that time. Until then I'm worried that this field does nothing and it's very hard to document how to populate a useless field. In particular if it's something other than the logical AND of the individual snapshots' fields, we'd be unable to explain why because nothing consumes the field.

Reply:
I think there is one more angle to that, that said, if SP want to do some 'pre work' before declaring that , `all good `  on the global level of group snapshot create request handling, the SP driver can make use of this field. It could be sometimes simple as -  update a journal/metadata about a map of snapshot IDs belong to this group snapshot or mapping of volumes to snapshots in a state system, or verify everything has reached to the desired state for this request...etc for SP's further operations on this group snapshot.  
Ideally , a consumer is not supposed to use,   some part of the result ( in this case individual snapshot)  of group snapshot create request or artifacts,  till the SP driver declares `ready_to_use` as `true`.  
imo, its not just `clone/restore` operation this field could depend upon, a `status/fetch` operation like get/list have to be triggered only when group level flag is set, Isnt it ?

Reply:
So I agree that in principle there could be some useful work to be done after every snapshot is "ready to use" individually but before the whole group snapshot is "ready to use", in which case you could have every snapshot report true and the top level group object report false.

The problem is that nobody will care because the group will never be consumed by a clone operation, and the clone operation is the only thing that waits on ready_to_use. The magic of the group snapshot operation is in the "cut" phase, not the "upload" phase, and the uploads can in principle be unsynchronzed. As soon as any given snapshots is uploaded (i.e. ready_to_use) it should be cloneable under this design. If that's not what we intend that we should actually be specifying that the individual snapshots should also not report ready_to_use until the whole thing is uploaded (assuming the SP has an upload phase at all)

Reply:
I'm having difficulty writing the paragraph on CreateVolumeGroupSnapshot if ready_to_use is removed from the group snapshot.
Currently it is as follows:
https://github.com/container-storage-interface/spec/pull/519/files#diff-bc6661da34ecae62fbe724bb93fd69b91a7f81143f2683a81163231de7e3b545R2839-R2841
Since VolumeGroupSnapshot is a resource created in CSI, I think we should have a parameter that indicate its final state. Without ready_to_use, a group snapshot may be created but with different states.  It may be successful and done, or it may be cut but post-processing failed.  We don't have a way to show that.

Reply:
But what does failed mean in that case? All you can do with a group snapshot that failed to create is delete it, but that's all you can do with a successful group snapshot too. The individual snapshots are cloneable whether the group operation failed or not, and all of the snapshots must have been cut at the same time, or the original Create call would not have returned successfully.

My thinking is that in the future this success/fail at the group level may matter for the purposes of some group-clone operation, then we should add this field at that time. As long as the only things we can do are clone the individual snapshots, then this field doesn't make sense.

I thought about it from the opposite perspective: assuming we keep the field, how do we document how to interpret it? I came up empty, because the CO behaves exactly the same in the true and false cases. That was the red flag that drove me to suggest removing the field (for now at least).

Reply:
I removed ready_to_use: https://github.com/container-storage-interface/spec/pull/519/commits/411868e000b2bdb0bb1bbae985d28cd5c32e5ada